### PR TITLE
OpInfo for `nn.functional.batch_norm`

### PR DESCRIPTION
**Pull request by krshrimali**
Addresses https://github.com/facebookresearch/functorch/issues/78 and https://github.com/pytorch/pytorch/issues/54261.

* There exists `torch.batch_norm` but it takes an extra arg: `cudnn_enabled` (not there in functional variant). This is passed from the functional variant to `torch.batch_norm` here: https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2282. `test_variant_consistency_jit` fails with an error: (when passed an alias)
    ```python
    File "/home/krshrimali/Documents/Projects/Quansight/pytorch/test/test_ops.py", line 457, in _test_consistency_helper
    variant_forward = variant(cloned,
    TypeError: batch_norm() missing 1 required positional arguments: "cudnn_enabled"
    ```
    * I'm not sure of a solution to this, as AFIK - there is no way to pass a lambda wrapper for an alias. Hence, I've skipped adding this as an alias there.
    * On second thought, is this even an alias?

cc: @mruberry @zou3519 @kshitij12345

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1069,6 +1069,34 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
...
+    )
+
+    def generator():
+        for input_shape, kwargs in cases:
+            # args: running mean, running var, weight and bias should necessarily be of shape: (C,)
+            channels = input_shape[1]
+            yield SampleInput(
+                make_arg(input_shape),
+                args = (
+                    make_arg(channels), make_arg(channels), make_arg(channels), make_arg(channels)
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:1093]: For `running_mean`, don't set the `requires_grad` to `True`.

Something like this:
https://github.com/pytorch/pytorch/blob/741accb11e3eed6c2eb8171b200a8adadf1616fb/torch/testing/_internal/common_methods_invocations.py#L2470-L2471

**Comment by kshitij12345** [comment]
> There exists torch.batch_norm but it takes an extra arg: cudnn_enabled (not there in functional variant). This is passed from the functional variant to torch.batch_norm here: https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2282. test_variant_consistency_jit fails with an error: (when passed an alias)

In that case, wouldn't `torch.batch_norm` be the main implementation and `nn.functional.batch_norm` be an alias for it? So OpInfo should be for `torch.batch_norm` and `nn.functional.batch_norm` should be in the alias list?

**Comment by krshrimali** [comment]
> > There exists torch.batch_norm but it takes an extra arg: cudnn_enabled (not there in functional variant). This is passed from the functional variant to torch.batch_norm here: https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2282. test_variant_consistency_jit fails with an error: (when passed an alias)
> 
> In that case, wouldn't `torch.batch_norm` be the main implementation and `nn.functional.batch_norm` be an alias for it? So OpInfo should be for `torch.batch_norm` and `nn.functional.batch_norm` should be in the alias list?

I don't think that will also work since the functional variant doesn't accept an arg of `cudnn_enabled`.

```python
ERROR: test_variant_consistency_eager_batch_norm_cpu_float32 (__main__.TestCommonCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 373, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 780, in test_wrapper
    return test(*args, **kwargs)
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/test/test_ops.py", line 468, in test_variant_consistency_eager
    _test_consistency_helper(samples, variants)
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/test/test_ops.py", line 457, in _test_consistency_helper
    variant_forward = variant(cloned,
TypeError: batch_norm() got an unexpected keyword argument 'cudnn_enabled'
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:1093]: Thanks for the suggestion, @kshitij12345. This does help in getting rid of the error, and there are now real grad failures. Will take a look and see what's the issue. Thanks!

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Fixed now, `running_mean` and `running_var` needed to be of value > 0. Thanks, @kshitij12345 for the help!

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -6781,6 +6814,17 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
            )),
+    OpInfo('nn.functional.batch_norm',
+           aten_name='batch_norm',
+           dtypes=floating_types(),
+           dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
+           supports_out=False,
+           skips=(
+               # RuntimeError: deepEquals(input.iValue, deepCopiedInput) INTERNAL ASSERT FAILED at
+               # "../torch/csrc/jit/passes/utils/check_alias_annotation.cpp":142, please report a bug to PyTorch
+               SkipInfo('TestJit', 'test_variant_consistency_jit'),
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:6825]: Unsure about this test failure.

cc: @eellison

File: test/test_nn.py [hunk 3]
```diff
@@ -15782,26 +15782,6 @@ def conv2d_depthwise(x, weight):
...
-        for training in [False, True]:
-            _assertGradAndGradgradChecks(self, F.batch_norm, (input, running_mean, running_var, weight, bias,
-                                                              training, 0.1, 0.0001))
-
-    def test_batchnorm_grad(self, device):
-        self._test_batchnorm_grad(device)
-
-        if self.device_type == 'cuda' and self.has_cudnn():
-            with torch.backends.cudnn.flags(enabled=False):
-                self._test_batchnorm_grad(device)
```
Comment (zou3519) [inline test/test_nn.py:15802]: We still want to test that batch norm on CUDA without CUDNN still works. Is there a way to do that using OpInfo?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -1069,6 +1069,39 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
         return inputs
 
 
+def sample_inputs_batch_norm(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+    make_arg_without_grad = partial(make_tensor, device=device, dtype=dtype, requires_grad=False)
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:1074]: nit: "make arg without grad" sounds like "we're making an argument without a grad tensor", which confused me. We should rename it to something like "make_non_differentiable_arg" or "make_arg_without_requires_grad"

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -1069,6 +1069,39 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
...
+                    make_arg_without_grad(channels, low=0),  # running mean
+                    make_arg_without_grad(channels, low=0),  # running var
+                    make_arg(channels),  # weight
+                    make_arg(channels)   # bias
+                ),
+                kwargs=kwargs
+            )
+        # Test case for no optional kwargs
+        yield SampleInput(make_arg((1, 2, 3)), args=(make_arg_without_grad(2, low=0),
+                                                     make_arg_without_grad(2, low=0)))
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:1102]: It's not well documented, but it turns out `running_mean` and `running_var` are allowed to be None. Could we add a test case for that as well? (I'm not sure if they both have to be none or if it is fine for one of them to be not-none, we should run the code and see...)

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Fixed. Sorry for the confusion :)

File: test/test_nn.py (hunk 3 above)
Comment (krshrimali) [reply test/test_nn.py]: Having another OpInfo with a custom wrapper for the op can be one way, but not sure if it's preferred. Temporarily, I've brought the tests back. Do you think having a separate OpInfo for this variant makes sense?

Apologies for not focusing on this detail.

**Comment by zou3519** [comment]
> There exists torch.batch_norm

IMO torch.batch_norm should not exist and we just forgot to exclude it and now we can't because of a deprecation cycle... but I wouldn't worry about the additional "cudnn_enabled" argument

File: test/test_nn.py (hunk 3 above)
Comment (zou3519) [reply test/test_nn.py]: > Apologies for not focusing on this detail.

Comment (zou3519) [reply test/test_nn.py]: > Having another OpInfo with a custom wrapper for the op can be one way, but not sure if it's preferred. Temporarily, I've brought the tests back. Do you think having a separate OpInfo for this variant makes sense?

@mruberry what say you? I'm down to have a separate OpInfo for this variant (the "definitely not cudnn" variant). The weird thing is that sometimes the cudnn variant and the non-cudnn variant will do the same thing (if we're running on CPU for example, or if a machine just doesn't have cudnn I don't think there is a way to force it)

Comment (mruberry) [reply test/test_nn.py]: The ideal solution here is probably to add an OpInfo that does something like always applying the following two decorators to its tests:

- @onlyCUDA (so CPU tests are skipped)
- a decorator that disables cuDNN for the duration of the test

A wrapper that disables cuDNN during the call the op is probably OK and could replace the proposed decorator, too.

**Review by zou3519 — COMMENTED** [review]
test failures seem real

**Comment by krshrimali** [comment]
Sorry for the delay on this one. I've updated the PR to fix minor failures. @zou3519 - can you please take a look whenever you find the time? Thanks! :)

**Comment by krshrimali** [comment]
Have fixed the merge conflicts, @zou3519 - could you please take a look whenever you find the time? Thanks! :)

**Comment by zou3519** [comment]
> Have fixed the merge conflicts, @zou3519 - could you please take a look whenever you find the time? Thanks! :)

yes, will do! Sorry for the delayed response.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -7595,6 +7631,33 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+    OpInfo('nn.functional.batch_norm',
+           variant_test_name='without_cudnn',
+           aten_name='batch_norm',
+           dtypesIfCPU=empty_types(),
+           dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
+           supports_out=False,
+           skips=(
+               # RuntimeError: deepEquals(input.iValue, deepCopiedInput) INTERNAL ASSERT FAILED at
+               # "../torch/csrc/jit/passes/utils/check_alias_annotation.cpp":142, please report a bug to PyTorch
+               SkipInfo('TestJit', 'test_variant_consistency_jit'),
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:7657]: Please replace the SkipInfos (we've deleted those): https://github.com/pytorch/pytorch/pull/64972

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Done, apologies for missing this change. I've not tested the change locally, but the tests should pass with the changes.

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Update: fixed a typo (sorry for that). Tests should pass now. (they pass on local)

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -1259,6 +1260,41 @@ def generator():
...
+
+    # Ordered as: input shape, kwargs for training, momentum, eps
+    cases: Tuple[Tuple[int], dict] = (  # type: ignore[assignment]
+        ((S, S, S), {'training': True, 'momentum': 0.5, 'eps': 0.6}),
+        ((3, 2, 4), {'training': False, 'momentum': -1.2}),
+        ((3, 1), {'training': True, 'momentum': 0.0}),
+        ((2, 1), {'training': False}),
+        ((3, 2, 3, 4), {'training': True, 'momentum': -1.0, 'eps': 0.5}),
+        ((3, 2, 3, 4), {'training': False, 'momentum': -1.0, 'eps': 0.5}),
+        ((2, 1), {})
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:1275]: Does batch_norm support a batch dimension of size 0? (I'm not sure, so asking here)

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -1259,6 +1260,41 @@ def generator():
...
+                    make_arg_without_requires_grad(channels, low=0),  # running mean
+                    make_arg_without_requires_grad(channels, low=0),  # running var
+                    make_arg(channels),  # weight
+                    make_arg(channels)   # bias
+                ),
+                kwargs=kwargs
+            )
+        # Test case for no optional kwargs
+        # running_mean and running_var are required in evaluation mode (training: False) but not in training mode
+        yield SampleInput(make_arg((1, 2, 3)), args=(None, None), kwargs={'training': True})
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:1294]: Hmm, It turns out that weight and bias can also be None. Would it be worth testing the following additional cases?
- input, running_mean, running_var, weight, None, training=True
- input, running_mean, running_var, None, bias, training=False
- input, running_mean, running_var, None, None, training=False

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -7767,6 +7803,33 @@ def wrapper(x: np.ndarray, *args, **kwargs):
...
+           dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
+           supports_out=False,
+           skips=(
+               # RuntimeError: deepEquals(input.iValue, deepCopiedInput) INTERNAL ASSERT FAILED at
+               # "../torch/csrc/jit/passes/utils/check_alias_annotation.cpp":142, please report a bug to PyTorch
+               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),
+           ),
+           sample_inputs_func=sample_inputs_batch_norm),
+    # This variant tests batch_norm with cuDNN disabled on only CUDA devices
+    OpInfo('nn.functional.batch_norm',
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:7820]: I'm not completely sure that OpInfo should be in the business of testing with cuDNN disabled -- this seems very backend specific. Feel free to leave this or remove it, we can always figure out what the relationship is between OpInfos and testing with/without cudnn is later.

**Review by zou3519 — COMMENTED** [review]
This looks great! I added some minor suggestions for cases, please let me know what you think

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:1275]: Nice point, thanks for asking! Looks like it does:

```python
In [5]: torch.nn.functional.batch_norm(torch.randn(0,), None, None, training=True)
Out[5]: tensor([])

In [5]: torch.nn.functional.batch_norm(torch.randn(0,), None, None, training=True)
Out[5]: tensor([])
```

Will add a case for 0d tensor.

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: For 0-d tensors, with `weight` and `bias` as non-None values, it fails with following error:

```python
In [3]: torch.nn.functional.batch_norm(x, x, x, weight=None, bias=x, training=True)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-3-dfc6402a839f> in <module>
----> 1 torch.nn.functional.batch_norm(x, x, x, weight=None, bias=x, training=True)

~/anaconda3/lib/python3.8/site-packages/torch/nn/functional.py in batch_norm(input, running_mean, running_var, weight, bias, training, momentum, eps)
   2279         _verify_batch_size(input.size())
   2280
-> 2281     return torch.batch_norm(
   2282         input, weight, bias, running_mean, running_var, training, momentum, eps, torch.backends.cudnn.enabled
   2283     )

IndexError: select(): index 0 out of range for tensor of size [0] at dimension 0
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 8 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:1318]: Agreed, good suggestion. I've added these cases.

File: torch/testing/_internal/common_methods_invocations.py (hunk 9 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Interesting, I'll leave it for now if that's fine? I'm curious to see if any other OpInfo also needs to disable cuDNN for some tests, might be a good weekend investigation.

Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py]: Yup, let's leave this for now

**Review by zou3519 — APPROVED** [review]

**Comment by zou3519** [comment]
Test failures are unrelated; Attempting to merge

**Pull request closed without merging**